### ISSUE TITLE
Input/Hotkeys are locked in QWERTY

### ISSUE THREAD

**Issue by Mr-Figs** [libretro/RetroArch#10249]
### Description
When binding/rebinding inputs under the 'Input' menu, Retroarch behaves as if the keyboard is in the QWERTY layout, even if it is set to a different layout in the OS. Searching (with `/`) is not affected.

For example, 'Fullscreen Toggle' is bound to F by default. Pressing my F key doesn't do anything since it is mapped to my right hand in my layout. Pressing the T key works, however, since my T is where QWERTY's F is (left hand, index finger, home row).

I can manually bind fullscreen to my F key, but the menu shows it bound to U instead, since QWERTY's U is where my F is. This is a reasonable workaround but would make rebinding confusing. It also initializes all the shortcuts to the wrong keys by default. (This is debatable--control keys [z,x,c etc] are chosen based on position, so should depend on physical position, not legend, but most other keys have mnemonics [f for **f**ullscreen, m for **m**ute etc], so should map to a specific letter).

Tested with a default config. To reproduce, set a non-QWERTY layout and bind some keys to test one-by-one; the top row will spell out QWERTY no matter what.

### Bisect

Highly likely related to #8303. This used to work, but I hadn't used RA on a machine with a non-QWERTY layout in a while. I don't have a build environment set up but I could if it's necessary.

### Environment
- RetroArch: 1.8.4 9e25601
- OS: Kubuntu 19.10

**Comment by keithbowes** [comment]
Yeah, I had to look down on my keyboard to get the physical key I press.  So under Dvorak, I have fullscreen bound to y, as Dvorak f is where QWERTY y is.  Back when I used X, the X driver would handle it correctly, but in Wayland and KMS (and X with the suggested udev driver), it doesn't work and, as you say, input_keyboard_layout only works with search (and other times when you're presented with an input field).  The keys in my config being completely different than the logical keys is something I've been living with since I first used it (1.7.3).

[I assume that you're using Wayland, as KDE has pretty much deprecated the X11 backend.]

**Comment by Mr-Figs** [comment]
Actually, I'm still on X11. It's the default setting in Kubuntu, I didn't do anything special to change it back.

Since you mentioned, I tried out the different input drivers. Running in a window (X environment): sdl2, x, linuxraw, and null all have identical behavior, and all have the issue. The udev driver doesn't do anything, so I have to edit the config file to change it back.

In KMS mode (launched from a TTY without x running), by default, RA softlocks my system since the x driver doesn't work without x. RA blocks ctrl-alt-fn TTY switching and SysRq so the only solution is to hard-boot and edit the config file. This is a separate usability issue. The same thing happens with udev and sdl2.

Running in KMS mode with the null input driver causes a segfault on launch. Relevant log line:
`[INFO] [Video]: Graphics driver did not initialize an input driver. Attempting to pick a suitable driver.`
Again, separate issue. This one is minor.

With the linuxraw driver, KMS mode works. However, the input behavior is very different. Instead of navigating with enter and backspace, I have to navigate with Z and X. Toggling "Unified menu controls" has no effect. More importantly, rebinding is impossible. I can select a bind to edit, which brings up the dialog, but none of my inputs register. Pressing `/` to search has no effect.

So there's a host of weird KMS-related bugs that only show on my machine. On my system (using X), all the working input drivers behave the same.

**Comment by ntninja** [comment]
Based on the (reasonable) explanations in #8303, let's try to rephrase this into some actionable items:

(I did not include any mention of the KMS/DRM issues in the below as it appears to me to be another issue entirely.)

### The human-facing display of keys in the settings is confusing

The issue here is that RetroArch internally uses keycodes/scancodes, but instead of displaying the matching localized keysym/character in the keyboard settings will display the scancode's ANSI interpretation instead.

*Proposed Solution*: Map keycodes to their current localized representation before displaying them in the settings.

### The keys in the configuration file are equally confusing

Here things are not so obvious, as a conscious decision will first have to be made whether the store key configuration in fact refers to the to the *characters/symbols* of the current keyboard layout (what is printed on the keys, keysyms) or the physical *location* of the key on the keyboard (scancodes).

*Proposed solution*: After deciding on the above question, either do nothing (keysyms) or convert all existing ANSI keysyms to their generating scancodes and keep storing scancodes in the configuration from now in.
